### PR TITLE
add `mc mv` command

### LINKED ISSUE #3077: Support for mc mv
A customer requested `mc mv` command such that as it copies the files it removes from the source so that disk space is conserved. Downside of breaking it into `mc cp` followed by `mc rm` is that it would need double the disk usage (since `mc rm` runs only after `mc cp` finishes)

### PR DESCRIPTION
**Pull request by balamurugana**
Fixes #3077

### CODE REVIEW

File: cmd/cp-main.go [hunk 1]
```diff
@@ -211,14 +211,27 @@ func doCopy(ctx context.Context, cpURLs URLs, pg ProgressReader, encKeyDB map[st
 			TotalSize:  cpURLs.TotalSize,
 		})
 	}
-	return uploadSourceToTargetURL(ctx, cpURLs, pg, encKeyDB)
+
+	urls := uploadSourceToTargetURL(ctx, cpURLs, pg, encKeyDB)
+	if isMvCmd {
+		if urls.Error == nil {
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply cmd/cp-main.go:217]: done

**Comment by BigUstad** [comment]
I was testing with a bunch of files & 1 directory for mc mv and got the following bug:

Zamins-MacBook-Pro:mc-mv-3077 priyankraj$ mc mv --older-than 7d --newer-than 10d --recursive ~/Downloads/ prihome9000/bkt1
 0 B / ?  ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░▓ mc: <ERROR> Invalid argument ``. Invalid path, path cannot be empty
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x88 pc=0x478d751]

goroutine 40 [running]:
github.com/minio/mc/cmd.bgRemove.func1(0x0, 0x0, 0x0, 0x0)
	/Users/priyankraj/Documents/ws/src/github.com/mc-mv-3077/cmd/mv-main.go:246 +0x171
created by github.com/minio/mc/cmd.bgRemove
	/Users/priyankraj/Documents/ws/src/github.com/mc-mv-3077/cmd/mv-main.go:261 +0x9e

**Comment by balamurugana** [comment]
> I was testing with a bunch of files & 1 directory for mc mv and got the following bug:
> 
> Zamins-MacBook-Pro:mc-mv-3077 priyankraj$ mc mv --older-than 7d --newer-than 10d --recursive ~/Downloads/ prihome9000/bkt1
> 0 B / ? ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░▓ mc: Invalid argument ``. Invalid path, path cannot be empty
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x88 pc=0x478d751]
> 
> goroutine 40 [running]:
> github.com/minio/mc/cmd.bgRemove.func1(0x0, 0x0, 0x0, 0x0)
> /Users/priyankraj/Documents/ws/src/github.com/mc-mv-3077/cmd/mv-main.go:246 +0x171
> created by github.com/minio/mc/cmd.bgRemove
> /Users/priyankraj/Documents/ws/src/github.com/mc-mv-3077/cmd/mv-main.go:261 +0x9e

Fixed

File: cmd/mv-main.go [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+			session.Header.CommandArgs = ctx.Args()
+		}
+	}
+
+	e := doCopySession(ctx, session, encKeyDB, true)
+	if session != nil {
+		session.Delete()
+	}
+
+	console.Colorize("Copy", "Waiting move operations to complete")
```
Comment (krisis) [inline cmd/mv-main.go:352]:
```suggestion
	console.Colorize("Copy", "Waiting for move operations to complete")
```

File: cmd/mv-main.go [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+	rm.wg.Wait()
+}
+
+var rmManager = &removeManager{
+	removeMap: make(map[string]*removeClientInfo),
+	doneCh:    make(chan struct{}),
+}
+
+func bgRemove(url string) {
```
Comment (krisis) [inline cmd/mv-main.go:234]: It would be useful to have comments explaining why we use a different approach when `targetAlias` is empty .

Comment (balamurugana) [reply cmd/mv-main.go:234]: done

File: cmd/mv-main.go (hunk 2 above)
Comment (balamurugana) [reply cmd/mv-main.go]: done

File: cmd/mv-main.go [hunk 4]
```diff
@@ -0,0 +1,356 @@
+/*
+ * MinIO Client (C) 2014-2019 MinIO, Inc.
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply cmd/mv-main.go]: done

File: cmd/cp-url-syntax.go [hunk 5]
```diff
@@ -146,11 +154,17 @@ func checkCopySyntaxTypeC(srcURLs []string, tgtURL string, isRecursive bool, key
...
 			if !isRecursive {
+				if isMvCmd {
+					fatalIf(errInvalidArgument().Trace(srcURL), "To move a folder requires --recursive flag.")
+				}
 				fatalIf(errInvalidArgument().Trace(srcURL), "To copy a folder requires --recursive flag.")
 			}
 
 			// Check if we are going to copy a directory into itself
 			if isURLContains(srcURL, tgtURL, string(c.GetURL().Separator)) {
+				if isMvCmd {
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply cmd/cp-url-syntax.go:175]: done

File: cmd/cp-url-syntax.go [hunk 6]
```diff
@@ -146,11 +154,17 @@ func checkCopySyntaxTypeC(srcURLs []string, tgtURL string, isRecursive bool, key
 		if srcContent.Type.IsDir() {
 			// Require --recursive flag if we are copying a directory
 			if !isRecursive {
+				if isMvCmd {
+					fatalIf(errInvalidArgument().Trace(srcURL), "To move a folder requires --recursive flag.")
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply cmd/cp-url-syntax.go]: done

**Comment by balamurugana** [comment]
> Can we also add --disable-multipart flag support for mv as well?

done

**Pull request merged**